### PR TITLE
advanced cleanup mechanisms for setup first

### LINKED ISSUE #2586: [enhancement]: Further cleanup on setup first
### No existing issues.

- [X] There is no existing issue for my request.

### Description

On setup first we already have some cleanup processes, like file syncing ensuring sets are their etc. 

We should also ensure, that some garbage objects which were created sometimes in the past are deleted. Like `undefined` or `null` object on top-level which some adapters created accidentally and never deleted. Checking that `system.adapter.XY` objects are deleted if the adapter is no longer installed (check that adapter object is assigned to this host, if yes and adapter not installed and no instance -> cleanup)

Check if there are objects which contain forbidden chars, if so delete them as interaction is no longer possible with them, it won't cause any damage. 

### Why?

we recently had a strange bug https://github.com/ioBroker/ioBroker.js-controller/issues/2579 where a top level `null` object caused issues. Of course this was due to another fixable issue, but the anger for users could have been avoided.

Another thing was, that many user had an `system.adapter.vis` object but no vis was installed anymore, this lead to the bahvior that controller tries to collect vis datapoints, but no vis folder etc were existing anymore.

### How?

Simple DB interactions on setup first

### PR DESCRIPTION
**Pull request by foxriver76**
- will close #2586
- cleaning adapter objects of already removed adapters
- for this we create adapters now also in `system.host.XY.adapters.adapterName`, these adapters are the single point of truth for each host now on multihost systems
- we clean up objects and states which have invalid IDs
- we clean up so-called garbage objects, these are objects where we know they were accidentaly created by some adapters or get created again from time to time, e.g. a dev does not perform enough checks and creates an object with the id `undefined`. (note here we could also prevent this already on `setObject` methods, checking okay it is a string but it has a garbage id do not create..)

### CODE REVIEW

File: packages/cli/src/lib/setup/setupSetup.ts [hunk 1]
```diff
@@ -1024,6 +1022,214 @@ Please DO NOT copy files manually into ioBroker storage directories!`
...
+
+        if (!this.states) {
+            throw new Error('States not set up, call setupObjects first');
+        }
+
+        const objIds = await this.objects.getKeys('*');
+
+        if (objIds) {
+            for (const id of objIds) {
+                if (FORBIDDEN_CHARS.test(id)) {
```
Comment (Apollon77) [inline packages/cli/src/lib/setup/setupSetup.ts:1196]: wikl that work and not throw? Do you might need the special mpde as option for that that ignores these character errors?

**Comment by Apollon77** [comment]
PS: readme adjustment maybe needed?

File: packages/cli/src/lib/setup/setupSetup.ts (hunk 1 above)
Comment (foxriver76) [reply packages/cli/src/lib/setup/setupSetup.ts:1196]: no on db level there is no check

File: packages/cli/src/lib/setup/setupSetup.ts [hunk 2]
```diff
@@ -1024,6 +1022,214 @@ Please DO NOT copy files manually into ioBroker storage directories!`
...
+                }
+            }
+        }
+
+        const stateIds = await this.states.getKeys('*');
+
+        if (stateIds) {
+            for (const id of stateIds) {
+                if (FORBIDDEN_CHARS.test(id)) {
+                    await this.states.delState(id);
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/cli/src/lib/setup/setupSetup.ts:1208]: no on db level there is no check this works as wanted

File: packages/db-states-file/src/lib/states/statesInMemFileDB.js [hunk 3]
```diff
@@ -51,18 +51,19 @@ class StatesInMemoryFileDB extends InMemoryFileDB {
         };
         super(settings);
 
+        this.META_ID = '**META**';
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/db-states-file/src/lib/states/statesInMemFileDB.js:54]: I am personally a fan of this if only needed in this class. Else we would export it somewhere central if used in multiple places.

**Comment by foxriver76** [comment]
> PS: readme adjustment maybe needed?

Do you mean adding a section for the cleanup mechanic? It is not a specific feature someone would need to be aware of but we can add if you think it is super useful but maybe does not match our current readme structure.

**Comment by Apollon77** [comment]
readme: No i mean document the new object location :-)

**Comment by foxriver76** [comment]
> readme: No i mean document the new object location :-)

Ah you mean the additional `adapter` object at `system.host.adapter.XY`?

**Comment by Apollon77** [comment]
... and also that the old states stay duplicated and such

**Review by Apollon77 — APPROVED** [review]

**Pull request merged**